controllers/pod.go

```diff
@@ -185,155 +160,227 @@ func (r *SingleClusterReconciler) quickRestart(
+			failedPods = append(failedPods, pod)
 			continue
 		}
+		activePods = append(activePods, pod)
+	}
+	return failedPods, activePods
+}
+
+func (r *SingleClusterReconciler) safelyDeletePodsAndEnsureImageUpdated(
+	rackState RackState, podsToUpdate []*corev1.Pod, ignorablePods []corev1.Pod,
+) reconcileResult {
 
-		r.Log.Info("Pod is restarted", "podName", updatedPod.Name)
-		started = true
-		break
+	failedPods, activePods := getFailedAndActivePods(podsToUpdate)
+
+	// If already dead node (failed pod) then no need to check node safety, migration
+	if len(failedPods) != 0 {
+		r.Log.Info("Restart failed pods with updated container image", "pods", getPodNames(failedPods))
+		if res := r.deletePodAndEnsureImageUpdated(rackState, failedPods); !res.isSuccess {
+			return res
+		}
 	}
 
-	// TODO: In what situation this can happen?
-	if !started {
-		r.Log.Error(
-			err, "Pod is not running or ready. Pod might also be terminating",
-			"podName", pod.Name, "status", pod.Status.Phase,
-			"DeletionTimestamp", pod.DeletionTimestamp,
-		)
-	} else {
-		r.Recorder.Eventf(r.aeroCluster, corev1.EventTypeNormal, "PodRestarted",
-			"[rack-%s] Restarted Pod %s", pod.Labels[asdbv1beta1.AerospikeRackIdLabel], pod.Name)
+	if len(activePods) != 0 {
+		r.Log.Info("Restart active pods with updated container image", "pods", getPodNames(failedPods))
+		if res := r.waitForMultipleNodesSafeStopReady(activePods, ignorablePods); !res.isSuccess {
+			return res
+		}
+		if res := r.deletePodAndEnsureImageUpdated(rackState, activePods); !res.isSuccess {
+			return res
+		}
 	}
 
 	return reconcileSuccess()
 }
 
 func (r *SingleClusterReconciler) deletePodAndEnsureImageUpdated(
-	rackState RackState, p corev1.Pod, ignorablePods []corev1.Pod,
-) reconcileResult {
-	// If already dead node, so no need to check node safety, migration
-	if err := utils.CheckPodFailed(&p); err == nil {
-		if res := r.waitForNodeSafeStopReady(
-			&p, ignorablePods,
-		); !res.isSuccess {
-			return res
+	rackState RackState, podsToUpdate []*corev1.Pod) reconcileResult {
+
+	// Delete pods
+	for _, p := range podsToUpdate {
+		if err := r.Client.Delete(context.TODO(), p); err != nil {
+			return reconcileError(err)
 		}
-	}
+		r.Log.V(1).Info("Pod deleted", "podName", p.Name)
 
-	r.Log.V(1).Info("Delete the Pod", "podName", p.Name)
-	// Delete pod
-	if err := r.Client.Delete(context.TODO(), &p); err != nil {
-		return reconcileError(err)
+		r.Recorder.Eventf(r.aeroCluster, corev1.EventTypeNormal, "PodWaitUpdate",
+			"[rack-%d] Waiting to update Pod %s", rackState.Rack.ID, p.Name)
 	}
-	r.Log.V(1).Info("Pod deleted", "podName", p.Name)
 
-	r.Recorder.Eventf(r.aeroCluster, corev1.EventTypeNormal, "PodWaitUpdate",
-		"[rack-%d] Waiting to update Pod %s", rackState.Rack.ID, p.Name)
-	// Wait for pod to come up
+	return r.ensurePodsImageUpdated(podsToUpdate)
+}
+
+func (r *SingleClusterReconciler) ensurePodsImageUpdated(podsToCheck []*corev1.Pod) reconcileResult {
```

Comment:
ensurePodsImageUpdated and ensurePodsRunningAndReady seems similar with a check change. Do you think these can be merge and use that by passing an argument to differentiate?